ethexe/contracts/src/Mirror.sol

```diff
@@ -172,14 +176,105 @@ contract Mirror is IMirror {
 
     /// @dev Value never sent since goes to mailbox.
     function _sendMailboxedMessage(Gear.Message calldata _message) private {
+        bytes calldata payload = _message.payload;
+
+        if (_message.destination == ETH_EVENT_ADDR && _message.value == 0 && payload.length > 0) {
+            uint256 topicsLength;
+            assembly ("memory-safe") {
+                topicsLength := shr(248, calldataload(payload.offset))
+            }
+
+            if (topicsLength >= 1 && topicsLength <= 4) {
+                uint256 topicsLengthInBytes;
+                unchecked {
+                    topicsLengthInBytes = 1 + topicsLength * 32;
+                }
+
+                if (payload.length >= topicsLengthInBytes) {
+                    bytes32 topic1;
+                    assembly ("memory-safe") {
+                        topic1 := calldataload(add(payload.offset, 1))
+                    }
+
+                    if (
+                        topic1 != StateChanged.selector && topic1 != MessageQueueingRequested.selector
+                            && topic1 != ReplyQueueingRequested.selector && topic1 != ValueClaimingRequested.selector
+                            && topic1 != ExecutableBalanceTopUpRequested.selector && topic1 != Message.selector
+                            && topic1 != Reply.selector && topic1 != ValueClaimed.selector
+                    ) {
+                        uint256 size;
+                        unchecked {
+                            size = payload.length - topicsLengthInBytes;
+                        }
+
+                        uint256 memPtr = Memory.allocate(size);
+                        assembly ("memory-safe") {
+                            calldatacopy(memPtr, add(payload.offset, topicsLengthInBytes), size)
+                        }
+
+                        bytes32 topic2;
+                        bytes32 topic3;
+                        bytes32 topic4;
+                        assembly ("memory-safe") {
+                            topic2 := calldataload(add(payload.offset, 33))
```

Comment:
And here? What the magic numbers?))

Reply:
since we use `topicsLength` to get type of log: `log1`, `log2`, `log3`, `log4` we need to add `+ 1` to already existing offset (`+ 32`, `+ 64`, ...)